### PR TITLE
Add Doc Classification changes

### PR DESCRIPTION
**Pull request by Ashutosh-Adhikari**
These are the series of updates made before the NAACL submission, for the document classification project. 

@Impavidity please review those if possible for you.

### CODE REVIEW

File: common/train.py [hunk 1]
```diff
@@ -28,6 +28,7 @@ class TrainerFactory(object):
         'Reuters': ReutersTrainer,
         'AAPD': ReutersTrainer,
         'IMDB': ReutersTrainer,
+        'Yelp2014': ReutersTrainer,
```
Comment (daemon) [inline common/train.py]: Inconsistent name.

File: common/evaluation.py [hunk 2]
```diff
@@ -28,6 +28,7 @@ class EvaluatorFactory(object):
         'Reuters': ReutersEvaluator,
         'AAPD': ReutersEvaluator,
         'IMDB': ReutersEvaluator,
+        'Yelp2014': ReutersEvaluator,
```
Comment (daemon) [inline common/evaluation.py]: Here as well.

File: char_cnn/__main__.py [hunk 3]
```diff
@@ -73,13 +77,20 @@ def evaluate_dataset(split_name, dataset_cls, model, embedding, loader, batch_si
     random.seed(args.seed)
     logger = get_logger()
 
-    # Set up the data for training SST-1
-    if args.dataset == 'Reuters':
-        train_iter, dev_iter, test_iter = Reuters.iters(args.data_dir, args.word_vectors_file, args.word_vectors_dir, batch_size=args.batch_size, device=args.gpu, unk_init=UnknownWordVecCache.unk)
-    elif args.dataset == 'AAPD':
-        train_iter, dev_iter, test_iter = AAPD.iters(args.data_dir, args.word_vectors_file, args.word_vectors_dir, batch_size=args.batch_size, device=args.gpu, unk_init=UnknownWordVecCache.unk)
-    else:
+    dataset_map = {
```
Comment (daemon) [inline char_cnn/__main__.py]: This is good practice. We should fix the older code in a later commit.

File: char_cnn/__main__.py [hunk 4]
```diff
@@ -10,7 +10,10 @@
 from common.evaluation import EvaluatorFactory
 from common.train import TrainerFactory
 from datasets.aapd import AAPDCharQuantized as AAPD
+from datasets.yelp2014 import Yelp2014CharQuantized as Yelp2014
```
Comment (daemon) [inline char_cnn/__main__.py]: Nitpick: see PEP8 import order:

```
from copy import deepcopy
import logging
import random

from sklearn import metrics
import numpy as np
import torch
import torch.nn.functional as F

from char_cnn.args import get_args
from char_cnn.model import CharCNN
from common.evaluation import EvaluatorFactory
from common.train import TrainerFactory
from datasets.aapd import AAPDCharQuantized as AAPD
from datasets.yelp2014 import Yelp2014CharQuantized as Yelp2014
from datasets.imdb import IMDBCharQuantized as IMDB
from datasets.reuters import ReutersCharQuantized as Reuters

```

File: char_cnn/__main__.py [hunk 5]
```diff
@@ -45,13 +48,14 @@ def get_logger():
...
-def evaluate_dataset(split_name, dataset_cls, model, embedding, loader, batch_size, device):
+def evaluate_dataset(split_name, dataset_cls, model, embedding, loader, batch_size, device, single_label):
     saved_model_evaluator = EvaluatorFactory.get_evaluator(dataset_cls, model, embedding, loader, batch_size, device)
     saved_model_evaluator.ignore_lengths = True
+    saved_model_evaluator.single_label = single_label
     scores, metric_names = saved_model_evaluator.get_scores()
-    logger.info('Evaluation metrics for {}'.format(split_name))
-    logger.info('\t'.join([' '] + metric_names))
-    logger.info('\t'.join([split_name] + list(map(str, scores))))
+    print('Evaluation metrics for', split_name)
```
Comment (daemon) [inline char_cnn/__main__.py]: Is logger better here?

File: common/trainers/reuters_trainer.py [hunk 6]
```diff
@@ -85,9 +85,9 @@ def train(self, epochs):
         # model_outfile is actually a directory, using model_outfile to conform to Trainer naming convention
         os.makedirs(self.model_outfile, exist_ok=True)
         os.makedirs(os.path.join(self.model_outfile, self.train_loader.dataset.NAME), exist_ok=True)
-        print(header)
 
         for epoch in range(1, epochs + 1):
+            print('\n' + header)
```
Comment (daemon) [inline common/trainers/reuters_trainer.py]: Have we decided on 3.6? `f"\n{header}"` would be the way to do it.

File: datasets/yelp2014.py [hunk 7]
```diff
@@ -0,0 +1,89 @@
+import numpy as np
```
Comment (daemon) [inline datasets/yelp2014.py]: See import order as well here.

File: datasets/yelp2014.py [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+        :param batch_size: batch size
+        :param device: GPU device
+        :return:
+        """
+        train, val, test = cls.splits(path)
+        return BucketIterator.splits((train, val, test), batch_size=batch_size, repeat=False, shuffle=shuffle, device=device)
+
+
+class Yelp2014Hierarchical(Yelp2014):
+    In_FIELD = Field(batch_first=True, tokenize=clean_string)
```
Comment (daemon) [inline datasets/yelp2014.py]: `In_FIELD`?

File: han/__main__.py [hunk 9]
```diff
@@ -12,10 +12,13 @@
 from datasets.sst import SST2
 from datasets.reuters import ReutersHierarchical as Reuters
 from datasets.aapd import AAPDHierarchical as AAPD
+from datasets.imdb import IMDBHierarchical as IMDB
```
Comment (daemon) [inline han/__main__.py]: Import order.

File: han/args.py [hunk 10]
```diff
@@ -13,12 +13,12 @@ def get_args():
     parser.add_argument('--word_num_hidden', type = int, default = 50)
```
Comment (daemon) [inline han/args.py]: No space in kwarg `=`.

File: lstm_regularization/args.py [hunk 11]
```diff
@@ -6,22 +6,22 @@
 def get_args():
     parser = ArgumentParser(description="Regularized LSTM for text classification with Regularization")
     parser.add_argument('--no_cuda', action='store_false', help='do not use cuda', dest='cuda')
-    parser.add_argument('--gpu', type=int, default=0) # Use -1 for CPU
+    parser.add_argument('--gpu', type=int, default=0)  # Use -1 for CPU
```
Comment (daemon) [inline lstm_regularization/args.py]: Add `help="Use -1 for CPU"`?

File: common/train.py (hunk 1 above)
Comment (achyudhk) [reply common/train.py]: Can you please elaborate on this? Do you suggest we rename ReutersTrainer to GenericDocTrainer or something?

File: char_cnn/__main__.py (hunk 5 above)
Comment (achyudhk) [reply char_cnn/__main__.py]: We were redirecting to stdout in our run scripts, and using a logger didn't do that for us. We would end up with the results and warnings alone on the screen.

File: common/train.py (hunk 1 above)
Comment (daemon) [reply common/train.py]: Yeah sure.

File: char_cnn/__main__.py (hunk 5 above)
Comment (daemon) [reply char_cnn/__main__.py]: I see.

File: char_cnn/__main__.py (hunk 3 above)
Comment (Ashutosh-Adhikari) [reply char_cnn/__main__.py]: Cool. Thanks!

**Pull request merged**